chore: enables editing of an observation

cimigree opened the pull request:
closes #461 
Description:
- Adds a hook using tanstack for useCreatedByToDeviceId 
- Uses the hook in three screens/ components to convert the created by (wherever it happens to come from) to device Id
- Gets the device id from the useDeviceInfo hook in three observation files. Note, I could not use the 'react-native-device-info' deviceId as was mentioned in the issue, as that is the wrong format and instead used the one from our API
 - uses isPending instead of isLoading for both hook calls
 - fixed the typo in the ObservationListItem
 - Although it is not asked for, I used the deviceId and conversion of createdBy to check for isMine in the ObservationListItem for wherever it ends up being used. Not sure if this is desired and can remove it is not the right thing to do

Note:
These changes work well to edit an observation, but it leaves the button in place to change an observation category (preset) which comes from this file: src/frontend/screens/ObservationEdit/PresetView.tsx
If I attempt to change the preset, I get all kinds of errors. In this PR, should I make it so that Change is not there on an existing observation that is being edited? Or is that as expected? Or should it be a different issue/ PR?

Screen recording demoing editing:
<img src='https://github.com/user-attachments/assets/621ce0b1-a68c-4b1b-b8b9-ab812f710af5' width="33%" height="33%"/>

Screen Recording demoing error:
<img src='https://github.com/user-attachments/assets/823578fc-4422-455b-a985-51cc6c360cce' width="33%" height="33%"/>

achou11 commented on src/frontend/screens/ObservationsList/ObservationListItem.tsx:
```diff
@@ -33,20 +37,25 @@ function ObservationListItemNotMemoized({
       style={{flex: 1, height: 80}}>
       <View
-        style={[styles.container, style, !isMine && styles.syncedObservation]}>
+        style={[styles.container, style, isMine && styles.syncedObservation]}>
```
the `syncedObservation` style should be applied to observations from others, not your own

```suggestion
        style={[styles.container, style, !isMine && styles.syncedObservation]}>
```

achou11 commented on src/frontend/hooks/server/deviceInfo.ts:
```diff
@@ -35,6 +37,17 @@ export const useEditDeviceInfo = () => {
   });
 };
 
+export const useCreatedByToDeviceId = (createdBy: string) => {
+  const project = useActiveProject();
+
+  return useQuery({
+    queryKey: [CREATED_BY_TO_DEVICE_ID_KEY, createdBy],
+    queryFn: async () => {
+      return await project.$createdByToDeviceId(createdBy);
+    },
+  });
+};
```
nit: not sure if this belongs here, as device info is more about application-level information about a member's device. this is more about identity within a project, so i'd lean towards placing this in `server/projects`

cimigree replied:
Ok. I am ambivalent about it, so I moved it as you specified.

achou11 commented on src/frontend/screens/Observation/ObservationHeaderRight.tsx:
```diff
@@ -35,8 +41,9 @@ export const ObservationHeaderRight = ({
 
-  const canEdit =
-    observationWithPreset.observation.createdBy === data?.deviceId || false;
+  const {deviceId} = deviceInfo || {};
+  const canEdit = createdByDeviceId === deviceId || false;
```
nit

```suggestion
  const canEdit = createdByDeviceId === deviceInfo?.deviceId;
```

achou11 commented on src/frontend/screens/Observation/index.tsx:
```diff
@@ -57,9 +61,15 @@ export const ObservationScreen: NativeNavigationComponent<'Observation'> = ({
-  const deviceId = '';
   const {lat, lon, createdBy} = observation;
-  const isMine = deviceId === createdBy;
+  const {data: deviceInfo, isPending: isDeviceInfoPending} = useDeviceInfo();
+  const {deviceId} = deviceInfo || {};
+  const {data: convertedDeviceId, isPending: isCreatedByDeviceIdPending} =
+    useCreatedByToDeviceId(createdBy);
+  const isMine =
+    !isDeviceInfoPending &&
+    !isCreatedByDeviceIdPending &&
+    deviceId === convertedDeviceId;
```
nit: ideally this page would make use of the query statuses to wait before rendering relevant parts or passing data to components. for example, don't render the `ButtonFields` until data from both queries is available.

aside from that, i think this check can be more explicit to make sure that `deviceId` and `convertedDeviceId` are not undefined (at the TS level, at least):

```suggestion
  const {data: deviceInfo} = useDeviceInfo();
  const {data: convertedDeviceId} = useCreatedByToDeviceId(createdBy);
  const isMine =
    deviceInfo?.deviceId !== undefined &&
    convertedDeviceId !== undefined &&
    deviceInfo.deviceId === convertedDeviceId;
```

cimigree replied:
The only thing I found to conditionally render was the button fields, which I did. Is that what you had in mind? Were there other parts or components that were not being handled as you wished?

achou11 replied:
that was the only one that seemed relevant, thanks!

achou11 commented on src/frontend/screens/ObservationsList/ObservationListItem.tsx:
```diff
@@ -33,20 +37,25 @@ function ObservationListItemNotMemoized({
   const {preset} = useObservationWithPreset(observation.docId);
-  const deviceId = '';
+  const {data: deviceInfo} = useDeviceInfo();
+  const {deviceId} = deviceInfo || {};
```
nit: use optional chaining wherever `deviceId` is needed

```suggestion
```

achou11 reviewed (changes requested):
One blocking issue related to how items are rendered in the observation list screen

achou11 reviewed (commented):
Might be outside the scope of this PR but I'm noticing that once I enter the Edit Observation screen, the attachments do not load:

<img width="200" alt="image" src="https://github.com/user-attachments/assets/dc3bf5a0-e09a-4718-b653-9f1c22dcb5b7">

Seems to be an issue with how we're deriving photo draft state upon entering that screen ([link](https://github.com/digidem/comapeo-mobile/blob/de2d49e66d40e72350e1682c7cfd9c470ba918c1/src/frontend/hooks/persistedState/usePersistedDraftObservation/photosMethods.ts)). To put it shortly, we're [not keeping the blob uri for the photo](https://github.com/digidem/comapeo-mobile/blob/de2d49e66d40e72350e1682c7cfd9c470ba918c1/src/frontend/hooks/persistedState/usePersistedDraftObservation/photosMethods.ts#L52-L55) (see [`SavedPhoto`](https://github.com/digidem/comapeo-mobile/blob/de2d49e66d40e72350e1682c7cfd9c470ba918c1/src/frontend/hooks/persistedState/usePersistedDraftObservation/photosMethods.ts#L8-L14) type), which is the data that's eventually passed to that thumbnail previewer. without the uri, the thumbnail stays stuck in the loading state.

initial thought is to address that issue in this PR to be consistent with "not exposing broken features", but not sure how much work that would take. personally okay with it being a follow-up as long as it's addressed soon after.

achou11 commented:
> If I attempt to change the preset, I get all kinds of errors. In this PR, should I make it so that Change is not there on an existing observation that is being edited? Or is that as expected? Or should it be a different issue/ PR?

haven't tested this myself yet, but wondering if this is because that ObservationEdit screen is a little outdated in terms of implementation? Erik introduced a shared `Editor` component that we ideally use for this screen (was an eventually planned follow-up task) - maybe using that would fix the issue you're noticing?

cimigree commented on src/frontend/screens/Observation/index.tsx:
```diff
@@ -57,9 +60,15 @@ export const ObservationScreen: NativeNavigationComponent<'Observation'> = ({
   const {lat, lon, createdBy} = observation;
-  const isMine = deviceId === createdBy;
+  const {data: deviceInfo, isPending: isDeviceInfoPending} = useDeviceInfo();
+  let {data: convertedDeviceId, isPending: isCreatedByDeviceIdPending} =
```
Oops! Good catch! Leftover from some testing

cimigree commented:
- Fixed the let
- Fixed the rendering of that style that shouldn't be flickering
- Fixed (as best as I could figure out, very happy to take feedback) the rendering of photos in the ObservationEdit screen. 
Here it rendering instead of the loader
<img src="https://github.com/user-attachments/assets/fb434634-e76a-4b8c-ad7a-e886865918ed" width="33%" height="33%"/>

Note and frustration:
If I go to create an observation, take a photo, add details, and then save the details or go back to the Observation screen, then I get the loader. But that is not fixed if I navigate to the ObservationCreate Screen or do navigation.goBack() from the ObservationFields screen instead of navigating to the ObservationEdit screen. Ugh. Any chance I could leave that edge case for a different issue? Once you save it, it works fine.
